dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java

```diff
@@ -472,40 +598,38 @@ private boolean hasAttributeOptions( DimensionalItemObject o )
     }
 
     /**
-     * Evaluates the skip test expression for any sample periods in which skip
-     * test data occurs. For any combination of period and attribute option
-     * combo where the skip test is true, removes all sample data with that
-     * combination of period and attribute option combo.
+     * Finds sample periods that should be skipped based on the skip test.
      *
-     * @param sampleMap2 data values from sampled periods.
-     * @param skipTest the skip test expression.
-     * @param constantMap constants to use in skip expression if needed.
-     * @return the sample period map to use.
+     * @param allSamplePeriods all the sample periods.
+     * @param aocData data for this attribute option combo.
+     * @param skipTest the skip test.
+     * @param constantMap constants that may be in the skip expression.
+     * @return the sample periods to be skipped.
      */
-    private MapMap<Period, DimensionalItemObject, Double> applySkipTest(
-        MapMap<Period, DimensionalItemObject, Double> sampleMap2,
+    Set<Period> getSkippedPeriods( Set<Period> allSamplePeriods,
+        MapMap<Period, DimensionalItemObject, Double> aocData,
         Expression skipTest, Map<String, Constant> constantMap )
     {
+        Set<Period> skippedPeriods = new HashSet<>();
+
         if ( skipTest == null || StringUtils.isEmpty( skipTest.getExpression() ) )
         {
-            return sampleMap2;
+            return skippedPeriods;
         }
 
-        MapMap<Period, DimensionalItemObject, Double> periodValueMap = new MapMap<>();
-
-        for ( Period p : sampleMap2.keySet() )
+        for ( Period p : allSamplePeriods )
         {
-            if ( sampleMap2.get( p ) != null
-                && Boolean.TRUE != expressionService.getExpressionValue( skipTest.getExpression(),
-                    PREDICTOR_SKIP_TEST, sampleMap2.get( p ), constantMap, null,
+            if ( aocData.get( p ) != null
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [RC: Suspicious reference comparison of Boolean values (RC_REF_COMPARISON_BAD_PRACTICE_BOOLEAN)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rc-suspicious-reference-comparison-of-boolean-values-rc-ref-comparison-bad-practice-boolean).
This method compares two Boolean values using the == or != operator. Normally, there are only two Boolean values (Boolean.TRUE and Boolean.FALSE), but it is possible to create other Boolean objects using the new Boolean(b) constructor. It is best to avoid such objects, but if they do exist, then checking Boolean objects for equality using == or != will give results than are different than you would get using .equals(...).